### PR TITLE
Added "Principles" to explain the key technical principles behind Fastify

### PR DESCRIPTION
**Pull request by mcollina**
These principles are what @delvedor and myself wrote down a long time ago and presented at multiple conferences and events.

#### Checklist

- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

File: docs/Reference/Principles.md [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+9. Semantic versioning & Long Term Support;
+10. Specification adherence.
+
+## "Zero" Overhead in Production
+
+Fastify aims to implement its features by adding as minimal overhead to your application as possible.
+This is usually delivered by implementing fast algorithms and data structures, as well as JavaScript-specific
+features.
+
+Given that JavaScript does not offer zero-overhead data structures, ™his principle is at add with providing
```
Comment (delvedor) [inline docs/Reference/Principles.md:22]:
```suggestion
Given that JavaScript does not offer zero-overhead data structures, this principle is at add with providing
```

**Comment by Fdawgs** [comment]
@mcollina Mind sorting the line length for CI please?
The job annotations breaking up the lines are making it hard to review 😆 :
image

**Review by delvedor — APPROVED** [review]
LGTM!!

File: docs/Reference/Principles.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+1. “Zero” overhead in production;
+2. “Good” developer experience;
+3. Works great for small & big projects alike;
+4. Easy to migrate to microservices (or even serverless) and back;
+5. Security & data validation;
+6. If something could be a plugin, it likely should
+7. Easily testable;
+8. Do not monkeypatch core;
+9. Semantic versioning & Long Term Support;
+10. Specification adherence.
```
Comment (Fdawgs) [inline docs/Reference/Principles.md:15]:
```suggestion
1. “Zero” overhead in production
2. “Good” developer experience
3. Works great for small & big projects alike
4. Easy to migrate to microservices (or even serverless) and back
5. Security & data validation
6. If something could be a plugin, it likely should be
7. Easily testable
8. Do not monkeypatch core
9. Semantic versioning & Long Term Support
10. Specification adherence
```

File: docs/Reference/Principles.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+Given that JavaScript does not offer zero-overhead data structures, this principle
+is at add with providing a great developer experience and providing more features,
+as usually those cost some overhead.
+
+## "Good" Developer Experience
+
+Fastify aims to provide the best developer experience at the performance point
+it is operating.
+It provides a great out-of-the-box experience that it is flexible enough to be
+adapted to a various level of situations.
```
Comment (Fdawgs) [inline docs/Reference/Principles.md:33]:
```suggestion
It provides a great out-of-the-box experience that is flexible enough to be
adapted to a variety of situations.
```

File: docs/Reference/Principles.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+## Security and Data Validation
+
+Your web framework is the first point of contact with untrusted data, and it
+needs to act as the first line of defence for your system.
+
+## If something could be a plugin, it likely should
+
+We recognize that there are an infinite amount of use cases for an HTTP framework
+for Node.js. Catering to them in a single module would make the codebase unmaintainable.
+Thefore we provide hooks and options to allow you to customize the framework
```
Comment (Fdawgs) [inline docs/Reference/Principles.md:59]:
```suggestion
Therefore we provide hooks and options to allow you to customize the framework
```

File: docs/Reference/Principles.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+your codebase.
+
+## Easy to migrate to microservices (or even serverless) and back
+
+How you deploy your routes should not matter. The framework should "just work".
+
+## Security and Data Validation
+
+Your web framework is the first point of contact with untrusted data, and it
+needs to act as the first line of defence for your system.
```
Comment (Fdawgs) [inline docs/Reference/Principles.md:53]:
```suggestion
needs to act as the first line of defense for your system.
```
Even though it hurts my soul, the rest of the documentation uses American English.

American English: Defense
British English: Defence

File: docs/Reference/Principles.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+## If something could be a plugin, it likely should
+
+We recognize that there are an infinite amount of use cases for an HTTP framework
+for Node.js. Catering to them in a single module would make the codebase unmaintainable.
+Thefore we provide hooks and options to allow you to customize the framework
+as you please.
+
+## Easily testable
+
+Testing Fastify applications should be a first class concern.
```
Comment (Fdawgs) [inline docs/Reference/Principles.md:64]:
```suggestion
Testing Fastify applications should be a first-class concern.
```

File: docs/Reference/Principles.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+## "Zero" Overhead in Production
+
+Fastify aims to implement its features by adding as minimal overhead to your
+application as possible.
+This is usually delivered by implementing fast algorithms and data structures,
+as well as JavaScript-specific features.
+
+Given that JavaScript does not offer zero-overhead data structures, this principle
+is at add with providing a great developer experience and providing more features,
+as usually those cost some overhead.
```
Comment (Fdawgs) [inline docs/Reference/Principles.md:26]: Bit lost with this sentence as "is at add" doesn't make sense?

**Review by Fdawgs — CHANGES_REQUESTED** [review]

**Comment by Uzlopak** [comment]
I just put it into chat gpt: "Hi, please write the following in very hard english, with lots of foreign words and technical terms, so that is barely understandable: [...]"

But maybe sending it through chat gpt would make it better?

File: docs/Reference/Principles.md (hunk 7 above)
Comment (mcollina) [reply docs/Reference/Principles.md:26]:
```suggestion
Given that JavaScript does not offer zero-overhead data structures, this principle
is at odd with providing a great developer experience and providing more features,
as usually those cost some overhead.
```

Comment (mcollina) [reply docs/Reference/Principles.md:26]: add -> odd

**Review by Fdawgs — APPROVED** [review]

**Review by Uzlopak — APPROVED** [review]
LGTM

**Review by Eomm — COMMENTED** [review]
I would add a link to this file in the README

**Comment by gurgunday** [comment]
Thank you for solidifying these principles with this PR.

For me they are what make Fastify so unique and delightful to use ❤️

**Pull request merged**